### PR TITLE
core: Add support for interfaces through traits

### PR DESCRIPTION
**Pull request by math-fehr**
Compared to MLIR, we have both traits and interfaces as a single structure, which we call traits.

### CODE REVIEW

File: tests/test_traits.py [hunk 1]
```diff
@@ -170,3 +174,61 @@ class WrongTraitsType(IRDLOperation):
...
+class OpWithInterface(IRDLOperation):
+    name = "test.op_with_interface"
+    traits = frozenset([GetNumResultsTraitForOpWithOneResult()])
+
+    res: Annotated[OpResult, IntegerType]
+
+
+def test_interface():
+    """
+    Test the features of an trait with methods (An MLIR interface).
```
Comment (tobiasgrosser) [inline tests/test_traits.py:213]: an trait -> a trait

File: tests/test_traits.py [hunk 2]
```diff
@@ -78,10 +82,10 @@ def test_has_trait_object():
...
     Test the `has_trait` `Operation` method on a simple operation definition.
     """
-    assert TestOp.has_trait(LargerOperandTrait())
-    assert not TestOp.has_trait(LargerResultTrait())
-    assert not TestOp.has_trait(BitwidthSumLessThanTrait(0))
-    assert TestOp.has_trait(BitwidthSumLessThanTrait(64))
+    assert TestOp.has_trait(LargerOperandTrait)
+    assert not TestOp.has_trait(LargerResultTrait)
+    assert not TestOp.has_trait(BitwidthSumLessThanTrait, 0)
+    assert TestOp.has_trait(BitwidthSumLessThanTrait, 64)
```
Comment (AntonLydike) [inline tests/test_traits.py:88]: This makes it harder to check that two ops have the same trait, right? You can no longer to `op.has_trait(some_trait_var)`, instead you need to have a tuple `(type[OpTrait], Any)`. Not sure if this is the best way?

**Review by AntonLydike — APPROVED** [review]
Looks pretty good! I'm quite hyped to see this get going!

File: tests/test_traits.py (hunk 2 above)
Comment (math-fehr) [reply tests/test_traits.py:88]: You can get this with `op.get_traits_of_type(type(some_trait_var))` directly!

**Review by superlopuh — APPROVED** [review]

**Review by webmiche — APPROVED** [review]
From what I understand, this is essentially makes the interface an object that is part of the operation class, which then has the methods of the interface. This is in contrast to the classic definition of an interface where it the operation itself has these methods.

I think this is a worthwhile workaround, but I would not exactly claim that we have interface support with this. We have a way to hand a set of methods to several ops in one go, but the methods are not part of the API of the ops but hidden behind a layer of indirection.

Personally, I am fine with having this for the moment, but I think we will have to revisit the interface discussion at a later point.

**Comment by superlopuh** [comment]
Yep, I just re-read bits of the Toy tutorial, and I don't think this allows us to do most of the things discussed there, but that might be easier to do with concrete examples

**Comment by math-fehr** [comment]
> From what I understand, this is essentially makes the interface an object that is part of the operation class, which then has the methods of the interface. This is in contrast to the classic definition of an interface where it the operation itself has these methods.
> 
> I think this is a worthwhile workaround, but I would not exactly claim that we have interface support with this. We have a way to hand a set of methods to several ops in one go, but the methods are not part of the API of the ops but hidden behind a layer of indirection.
> 
> Personally, I am fine with having this for the moment, but I think we will have to revisit the interface discussion at a later point.

Yes, these interfaces are not directly in the operations, but on the object.
The reason why I do this is that I am scared that later on, people use the same names for some methods, and this would become impossible to fix in Python :/
Also, note that it is rare that you call these methods directly on the operations. It is way more likely that you query an interface on a given operation, and then call these methods.
If this becomes a problem, we can always change it, yeah!

**Comment by math-fehr** [comment]
> Yep, I just re-read bits of the Toy tutorial, and I don't think this allows us to do most of the things discussed there, but that might be easier to do with concrete examples

Do you have an example in mind?
I'm curious if I missed something. I would prefer to not push it if this PR has a fundamental flaw!

**Comment by superlopuh** [comment]
I may well be missing something, I think we might as well merge and iterate on design as we go along. My impression is that for things like inlining (https://mlir.llvm.org/docs/Tutorials/Toy/Ch-4/) we might have to go with an abstract base class approach instead, would be good to take a look together once I get to that stage

**Pull request merged**